src/main_nep/nep4.cu

```diff
@@ -258,62 +244,45 @@ static __global__ void apply_gnn_message_passing(
 {
   int n1 = threadIdx.x + blockIdx.x * blockDim.x;
   if (n1 < N) {
+    int t1 = g_type[n1];
     int num_neighbors_of_n1 = g_NN[n1];
-    const int F = ann.dim; // dimension of q_out, for now dim_out = dim_in.
-    for (int nu = 0; nu < F; nu++) {
-      float q_out_nu = g_messages[n1 + nu * N]; // fc(r_ii) = 1
-      // TODO perhaps normalize weights? Compare Kipf, Welling et al. (2016)
+    for (int nu = 0; nu < ann.dim; nu++) {
+      float q_out_nu = g_messages[n1 + nu * N];
       for (int j = 0; j < num_neighbors_of_n1; ++j) {
         int index = n1 + N * j;
         int n2 = g_NL[index];
+        int t2 = g_type[n2];
         float r12[3] = {g_x12[index], g_y12[index], g_z12[index]};
         float d12 = sqrt(r12[0] * r12[0] + r12[1] * r12[1] + r12[2] * r12[2]);
         float fc12;
         find_fc(para.rc_angular, para.rcinv_angular, d12, fc12);
+        fc12 *= ann.d[t1 * para.num_types + t2];
```

Comment:
and these are the parameters `dij` I suppose?

Reply:
Yes, these are the `num_types * num_types` `d_ij` parameters. I do not require the symmetry `d_ij = d_ji`.

Reply:
Why do you not have the symmetry? `fc12` is symmetric, and `d_ij` should also be symmetric right? Wouldn't the potential violate permutational invariance if not `d_ij =  d_ji`?